### ISSUE TITLE
No notification mail sent since upgrade 2.2 --> 2.4.2

### ISSUE THREAD

**Issue by guillaumefort** [naparuba/shinken#1733]
Hi,

My issue is concerning the notification system. I do not receive any mail since I upgraded to 2.4.2 (I was in 2.2). It used to work before that.

I saw the method changed (the plugin is now used by default), so I think you touched something about that in the code ?

The plugin is working: 

    shinken@monitor1:/etc/shinken$ /var/lib/shinken/libexec/notify_by_email.py -t -r mymail@address.com -f html
    2015-10-21 16:04:53,387:INFO: Mail sent successfuly

The config files are loaded:

    INFO: [Shinken] Processing object config file '/etc/shinken/notificationways/detailled-email.cfg'
    INFO: [Shinken] Processing object config file '/etc/shinken/notificationways/sms-android.cfg'
    INFO: [Shinken] Processing object config file '/etc/shinken/notificationways/email.cfg'

    shinken@monitor1:/etc/shinken$ cat /etc/shinken/notificationways/email.cfg

    define notificationway{
       notificationway_name            email
       service_notification_period     24x7
       host_notification_period        24x7
       service_notification_options    c,w,r
       host_notification_options       d,u,r,f,s
       service_notification_commands   notify-service-by-email ; send service notifications via email
       host_notification_commands      notify-host-by-email    ; send host notifications via email
    }

And the command (in commented lines, you can see what I also tried):

    shinken@monitor1:/etc/shinken$ cat /etc/shinken/commands/notify-service-by-email.cfg
    define command {
    command_name    notify-service-by-email
    command_line    $PLUGINSDIR$/notify_by_email.py -n service -S localhost -r $CONTACTEMAIL$ -f html -c "$NOTIFICATIONTYPE$,,$HOSTNAME$,,$HOSTADDRESS$,,$LONGDATETIME$" -o "$SERVICEDESC$,,$SERVICESTATE$,,$SERVICEOUTPUT$,,$SERVICEDURATION$"
    #command_line    /usr/bin/printf "%b" "Shinken Notification\n\nNotificatio Type: $NOTIFICATIONTYPE$\n\nService: $SERVICEDESC$\nHost: $HOSTNAME$\nAddress: $HOSTADDRESS$\nState: $SERVICESTATE$\n\nDate/Time: $DATE$ $TIME$\nAdditional Info : $SERVICEOUTPUT$\n" | /usr/bin/mail -s "** $NOTIFICATIONTYPE$ alert - $HOSTNAME$/$SERVICEDESC$ is $SERVICESTATE$ **" $CONTACTEMAIL$
    #command_line    /var/lib/shinken/libexec/notify_by_email.py -t -r mymail@address.com -f html 
    }

Despite all that, when the scheduler try to call the notification command, nothing happens:

    [1445433949] SERVICE NOTIFICATION: admin;backup;Memory Usage;WARNING;notify-service-by-email;MEMORY WARNING : Mem used: 80.46%, Swap used: 1.17%

I am stuck, I don't know where to investigate (when I enable the Debug mode in the scheduler or arbiter logs, nothing really clear appears to me).

Here are my system informations:

    shinken@monitor1:/etc/shinken$ lsb_release -a
    No LSB modules are available.
    Distributor ID:	Ubuntu
    Description:	Ubuntu 14.04.3 LTS
    Release:	14.04
    Codename:	trusty

    shinken@monitor1:/etc/shinken$ uname -a
    Linux monitor1 3.13.0-48-generic #80-Ubuntu SMP Thu Mar 12 11:16:15 UTC 2015 x86_64 x86_64 x86_64 GNU/Linux

Do you have any idea, can you tell me where to look ? Is there something in the system environment or in the way the functions are called that I am missing ?

Thank you !

**Comment by ianling** [comment]
What is the output of shinken --version?

**Comment by guillaumefort** [comment]
Hi,

Thank you for asking. I told it in the beginning of my post:

    shinken@monitor1:~$ shinken --version
    shinken 2.4.2

**Comment by olivierHa** [comment]
Hello,

Could you try an easy script like "echo $HOSTADDRESS$ >> /tmp/test" ?

Regards

Olivier


2015-10-22 10:10 GMT+02:00 Guillaume Fort <notifications@github.com>:

**Comment by ianling** [comment]
See this commit and check if your satellitelink.py file is up-to-date: https://github.com/naparuba/shinken/commit/fa90a46e7adedc93a944c787ef8364dd658762d7

This fixed my notifications issue a while back. I wonder if your install somehow didn't get the up-to-date version of this file.

**Comment by guillaumefort** [comment]
Thank you, I checked and my file is up-to-date.

I'm starting to think that I should reinstall from github repos... I dit it via "pip" and I don't see anything else to try :/

**Comment by geektophe** [comment]
As commands are invoked by a Popen like call, you cannot just use something like `echo $HOSTADDRESS$ >> /tmp/test`, which requires the command to be executed by a sh like shell.

Put this command in a script, and call the script by itself.

Could you also have a look at reactionner log file to check for errors ?

**Comment by guillaumefort** [comment]
Hi,

Ok, thank you, I did what you suggest and I made this script. I tried to execute it, and the /tmp/test file is populated (with a simple $, but populated). I modified the command line in the notification-service-by-email.cfg.

No rights problem for the script.

And... still nothing:

    [1445953677] SERVICE NOTIFICATION: admin;backup;BackupPC state;WARNING;notify-service-by-email;BACKUPPC WARNING - (1/4) failures

Nothing related to this in the reactionner log (in debug mode):

    [1445953677] DEBUG: [Shinken] Debug perf: ping [args:5.96046447754e-06] [aqu_lock:9.53674316406e-07][calling:2.14576721191e-06] [json:1.00135803223e-05] [global:1.90734863281e-05]
    [1445953677] DEBUG: [Shinken] Debug perf: get_external_commands [args:2.86102294922e-06] [aqu_lock:9.53674316406e-07][calling:1.59740447998e-05] [json:6.19888305664e-06] [global:2.59876251221e-05]
    [1445953678] DEBUG: [Shinken] Debug perf: ping [args:5.00679016113e-06] [aqu_lock:1.19209289551e-06][calling:1.90734863281e-06] [json:9.05990600586e-06] [global:1.71661376953e-05]
    [1445953678] DEBUG: [Shinken] Debug perf: get_external_commands [args:4.05311584473e-06] [aqu_lock:9.53674316406e-07][calling:1.69277191162e-05] [json:5.00679016113e-06] [global:2.69412994385e-05]
    [1445953679] DEBUG: [Shinken] Debug perf: ping [args:5.00679016113e-06] [aqu_lock:1.09672546387e-05][calling:2.14576721191e-06] [json:9.05990600586e-06] [global:2.71797180176e-05]
    [1445953679] DEBUG: [Shinken] Debug perf: get_external_commands [args:3.09944152832e-06] [aqu_lock:9.53674316406e-07][calling:1.81198120117e-05] [json:5.00679016113e-06] [global:2.71797180176e-05]

And of course, nothing in the file:
    shinken@monitor1:/etc/shinken$ cat /tmp/test 
    shinken@monitor1:/etc/shinken$

**Comment by geektophe** [comment]
The reactionner only logs errors, not successful commands. It is absolutely normal that there is nothing about your test inside it.

The log line in the scheduler log indicates that the command was successfully executed.

From what I can see, the shinken scheduling and execution part works as expected. The only thing I suspect is a change in the notification script parameters, or an incorrect command line parameter or expansion variable.

For your test file, it's absolutely normal that your test file only contains `$`, as the shell tries to expand `$HOSTADDRESS`, an undefined variable in your script context, followed by the `$` character.

To have the host address displayed in your test file, you have to let shinken evaluate `$HOSTADDRESS$`, and pass it as script parameter like this:

    define command {
        command_name    notify-service-by-email
        command_line    /path/to/your/script $HOSTADDRESS$
    }

and in `/path/to/your/script`

    #!/bin/bash
    echo $* >> /tmp/test

Could you try the following command with the script above ?

    define command {
        command_name    notify-service-by-email
        command_line    /path/to/your/script -n service -S localhost -r $CONTACTEMAIL$ -f html -c "$NOTIFICATIONTYPE$,,$HOSTNAME$,,$HOSTADDRESS$,,$LONGDATETIME$" -o "$SERVICEDESC$,,$SERVICESTATE$,,$SERVICEOUTPUT$,,$SERVICEDURATION$"
    }

You should have the full command line in your test file.

Next, try to execute /var/lib/shinken/libexec/notify_by_email.py with the parameters you got in the test file (don't forget to put the double quotes where they should be, they'll be lost by formatting).

**Comment by erny** [comment]
May this be related to: https://github.com/naparuba/shinken/issues/1752 ?

**Comment by arthurlogilab** [comment]
affected by this too.

**Comment by guillaumefort** [comment]
Hi,

Thank you for your answers. I have a little time to work on this case again so I tried what Geektophe said.

    shinken@monitor1:/etc/shinken$ cat /etc/shinken/commands/notify-service-by-email.cfg
    define command {
    command_name    notify-service-by-email
    command_line    /tmp/hep.bash -n service -S localhost -r $CONTACTEMAIL$ -f html -c "$NOTIFICATIONTYPE$,,$HOSTNAME$,,$HOSTADDRESS$,,$LONGDATETIME$" -o "$SERVICEDESC$,,$SERVICESTATE$,,$SERVICEOUTPUT$,,$SERVICEDURATION$"
    }

    shinken@monitor1:/etc/shinken$ cat /tmp/hep.bash 
    #!/bin/bash
    echo $* >> /tmp/test

    shinken@monitor1:/etc/shinken$ tail -f /var/log/shinken/schedulerd.log | grep -i notify
    [1448966816] SERVICE NOTIFICATION: admin;[snip];CRITICAL;notify-service-by-email;CRITICAL - [snip]

And... nothing in /tmp/, the test file is not created.

FYI, one of my first test before opening this case was to try the script the way you suggest it and it works.

    shinken@monitor1:/etc/shinken$ /var/lib/shinken/libexec/notify_by_email.py -n service -S localhost -r yeahyeah@yeahyeah.com -f html -c "NOTIFICATIONTYPE,,HOSTNAME,,HOSTADDRESS,,LONGDATETIME" -o "SERVICEDESC,,SERVICESTATE,,SERVICEOUTPUT,,SERVICEDURATION"
    2015-12-01 12:03:20,907:INFO: Mail sent successfuly

So far, Shinken core seems to call the command but it is simply not executed.

**Comment by guillaumefort** [comment]
I've finally found the solution...

I have cleaned everything and made a fresh install and everything worked fine (on the same instance) but from the moment I have restored my config files I faced the same problem again.

Ok, so there must be something in my conf that block email notification... I read again the shinken doc and especially the architecture diagram. And it jumped out at me ! I checked my reactionner configuration and I had the "manage_sub_realms" variable set to 0 while I had some realms configured...

I am shameful because I did not checked it sooner. The reason why it was at 0 was because my previous conf file was overwritten when I have upgraded with the last version and I did not restored it :/

Thank you for your help, at least I think I know a little how shinken works now !

I close the issue.
